google/cloud/forseti/services/inventory/cai_temporary_storage.py

```diff
@@ -0,0 +1,421 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Inventory temporary storage for Cloud Asset data."""
+import json
+import os
+import enum
+import tempfile
+
+from retrying import retry
+from sqlalchemy import Column
+from sqlalchemy import Enum
+from sqlalchemy import Index
+from sqlalchemy import PrimaryKeyConstraint
+from sqlalchemy import String
+from sqlalchemy import LargeBinary
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.pool import SingletonThreadPool
+
+from google.cloud.forseti.common.util import logger
+from google.cloud.forseti.services.dao import create_engine
+from google.cloud.forseti.services.inventory.base.gcp import AssetMetadata
+
+LOGGER = logger.get_logger(__name__)
+BASE = declarative_base()
+MAX_ALLOWED_PACKET = 32 * 1024 * 1024  # 32 Mb default mysql max packet size
+
+
+class ContentTypes(enum.Enum):
+    """Cloud Asset Inventory Content Types."""
+    resource = 1
+    iam_policy = 2
+
+
+SUPPORTED_CONTENT_TYPES = frozenset(item.name for item in list(ContentTypes))
+
+
+class CaiTemporaryStore(BASE):
+    """CAI temporary inventory table."""
+
+    __tablename__ = 'cai_temporary_store'
+
+    # Class members created in initialize() by mapper()
+    name = Column(String(512, collation='binary'), nullable=False)
+    parent_name = Column(String(255), nullable=True)
+    content_type = Column(Enum(ContentTypes), nullable=False)
+    asset_type = Column(String(255), nullable=False)
+    asset_data = Column(LargeBinary(length=(2**32) - 1), nullable=False)
+
+    __table_args__ = (
+        Index('idx_parent_name', 'parent_name'),
+        PrimaryKeyConstraint('content_type',
+                             'asset_type',
+                             'name',
+                             name='cai_temp_store_pk'))
+
+    # Assets with no parent resource.
+    UNPARENTED_ASSETS = frozenset([
+        'cloudresourcemanager.googleapis.com/Organization',
+        'cloudbilling.googleapis.com/BillingAccount',
+    ])
+
+    @classmethod
+    def from_json(cls, asset_json):
+        """Creates a database row object from the json data in a dump file.
+
+        Args:
+            asset_json (str): The json representation of an Asset.
+
+        Returns:
+            dict: database row dictionary or None if there is no data.
+        """
+        asset = json.loads(asset_json)
+        if len(asset['name']) > 512:
```

Comment:
Is this something we can resolve w/ the new approach, to increase the length of the asset name?

Reply:
Good call! Updated to 2048.